File: 프로그래머스/2/49994. 방문 길이/방문 길이.java
import java.util.*;

class Solution {
    private static int nx[] = {-1, 0, 1, 0};
    private static int ny[] = {0, -1, 0, 1};
    
    public int solution(String dirs) {
        int answer = 0;
        int len = dirs.length();
        
        boolean[][][] visited = new boolean[11][11][4];
        int x = 5, y = 5;
        
        for (int i=0; i<len; i++) {
            char c = dirs.charAt(i);
            int d = 0;
            if (dirs.charAt(i) == 'U') {
                d = 0;
            } else if (dirs.charAt(i) == 'L') {
                d = 1;
            } else if (dirs.charAt(i) == 'D') {
                d = 2;
            } else {
                d = 3;
            }
            
            int dx = x + nx[d];
            int dy = y + ny[d];
            
            if (dx >= 11 || dx < 0 || dy >= 11 || dy < 0) {
                continue;
            }
            
            if (!visited[dx][dy][d]) {
                answer += 1;
                visited[dx][dy][d] = true;
                d = (d%2 == 0) ? 2-d:4-d;
                visited[x][y][d] = true;
            }
            x = dx;
            y = dy;
        }
        
        return answer;
    }
}

File: 프로그래머스/2/49994. 방문 길이/방문 길이.py
from collections import deque
def solution(dirs):
    sets = set()
    d = {'U': (1, 0), 'D': (-1, 0), 'R': (0, 1), 'L': (0, -1)}
    x, y = 0, 0
    
    for di in dirs:
        dx, dy = d[di]
        nx = x + dx
        ny = y + dy
        
        if -5 <= nx <= 5 and -5 <= ny <= 5:
            sets.add(((x, y), (nx, ny)))
            sets.add(((nx, ny), (x, y)))
            x = nx
            y = ny
            
        
    return len(sets) // 2